internal/uvm/mem_allocator.go

```diff
@@ -0,0 +1,182 @@
+package uvm
```

Comment:
I think we should probably move this out to it's own package

Reply:
I feel like this definitely deserves a detailed comment on this file describing the implementation/why it's needed. If Kathryns suggestion is taken, then a detailed package comment would be amazing.